src/Solve.cpp

```diff
@@ -817,33 +756,44 @@ class SolveExpression : public IRMutator {
         if (op->name == var) {
             uses_var = true;
             return op;
-        } else if (const CacheEntry *e = scope.find(op->name)) {
+        }
+        if (const CacheEntry *e = scope.find(op->name)) {
             uses_var = uses_var || e->uses_var;
             failed = failed || e->failed;
             return e->expr;
-        } else if (const Expr *e = external_scope.find(op->name)) {
+        }
+        if (const Expr *e = external_scope.find(op->name)) {
             // Expressions in the external scope haven't been solved
             // yet. This will either pull its solution from the cache,
             // or solve it and then put it into the cache.
             return mutate(*e);
-        } else {
-            return op;
         }
+        return op;
     }
 
     Expr visit(const Let *op) override {
-        bool old_uses_var = uses_var;
-        bool old_failed = failed;
-        uses_var = false;
-        failed = false;
-        Expr value = mutate(op->value);
-        CacheEntry e = {value, uses_var, failed};
-        uses_var = old_uses_var;
-        failed = old_failed;
-
-        ScopedBinding<CacheEntry> bind(scope, op->name, e);
+        ScopedBinding<CacheEntry> bind(scope, op->name, mutate_inner(op->value));
         return mutate(op->body);
     }
+
+    Expr visit(const Select *op) override {
+        auto [true_value, true_uses_var, true_failed] = mutate_inner(op->true_value);
+        auto [false_value, false_uses_var, false_failed] = mutate_inner(op->false_value);
+
+        uses_var = uses_var || true_uses_var || false_uses_var;
+        failed = failed || true_failed || false_failed;
+
+        Expr condition = op->condition;
+
+        // select(cond, a, x) ~> select(!cond, x, a)
+        if (!true_uses_var && false_uses_var) {
+            condition = simplify(!condition);
+            std::swap(true_value, false_value);
+        }
+
+        condition = mutate(condition);
```

Comment:
I think this needs to track if condition uses the var and fail if it does and also true or false uses the var.